nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/controller/ControllerFacade.java

```diff
@@ -1853,6 +1859,62 @@ public SearchResultsDTO search(final String searchLiteral, final String activeGr
         return results;
     }
 
+    /**
+     * Get all user-defined data ingress ports provided by Listen Components (e.g., Processors and Controller Services)
+     *
+     * @param user the user performing the lookup
+     * @return the set of listen Ports accessible to the current user
+     */
+    public Set<ListenPortDTO> getListenPorts(final NiFiUser user) {
+
+        // Get all listen components for which the requesting user is authorized
+        final Set<ComponentNode> listenComponentNodes = flowController.getFlowManager().getAllListenComponents().stream()
+            .filter(componentNode -> componentNode.isAuthorized(authorizer, RequestAction.READ, user))
+            .collect(Collectors.toSet());
+
+        // If the current user doesn't have access to any listen components, return an empty result for ports
+        if (listenComponentNodes.isEmpty()) {
+            return Collections.emptySet();
+        }
+
+        // Now find all Listen Ports provided by the Listen Components. A listen component can provide multiple Listen Ports (e.g., ListenHTTP can have a data port and a health check port).
+        // The current Listen Ports for a component depend on configuration (e.g., port property value), so create a configuration context to provide ListenComponent.getListenPorts(context).
+        final Set<ListenPortDTO> listenPorts = new HashSet<>();
+        final ControllerServiceProvider controllerServiceProvider = flowController.getControllerServiceProvider();
+        for (final ComponentNode componentNode : listenComponentNodes) {
+            final ConfigurationContext configurationContext = new StandardConfigurationContext(componentNode, controllerServiceProvider, null);
+            final ConfigurableComponent component = componentNode.getComponent();
+            if (component instanceof ListenComponent listenComponent /* should always be true */) {
```

Comment:
For readability, recommend moving the comment above or below the line, instead of on the same line.